substrate/genesis-liquidity/pallet/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+          origin.clone().into(),
+          account,
+          Balance { coin: balance.coin, amount: Amount(coin_out) },
+        )?;
+        Coins::<T>::transfer(
+          origin.into(),
+          account,
+          Balance { coin: Coin::Serai, amount: Amount(sri) },
+        )?;
+
+        // return new amounts
+        (
+          (
+            user_shares.checked_sub(amount_to_remove).ok_or(Error::<T>::AmountOverflowed)?,
+            user_coins.checked_sub(coin_out).ok_or(Error::<T>::AmountOverflowed)?,
+          ),
+          (
+            supply.0.checked_sub(amount_to_remove).ok_or(Error::<T>::AmountOverflowed)?,
+            supply.1.checked_sub(coin_out).ok_or(Error::<T>::AmountOverflowed)?,
+          ),
+        )
+      } else {
+        if balance.amount.0 != GENESIS_LP_SHARES {
+          Err(Error::<T>::CanOnlyRemoveFullAmount)?;
+        }
+        let existing =
+          Liquidity::<T>::get(balance.coin, account).ok_or(Error::<T>::NotEnoughLiquidity)?;
+
+        // transfer to the user
+        Coins::<T>::transfer(
+          origin.into(),
+          account,
+          Balance { coin: balance.coin, amount: Amount(existing.1) },
+        )?;
+
+        (
+          (0, 0),
+          (
+            supply.0.checked_sub(existing.0).ok_or(Error::<T>::AmountOverflowed)?,
+            supply.1.checked_sub(existing.1).ok_or(Error::<T>::AmountOverflowed)?,
+          ),
+        )
+      };
+
+      // save
+      if new_shares.0 == 0 {
+        Liquidity::<T>::set(balance.coin, account, None);
+      } else {
+        Liquidity::<T>::set(balance.coin, account, Some(new_shares));
+      }
+      Supply::<T>::set(balance.coin, Some(new_supply));
+
+      Self::deposit_event(Event::GenesisLiquidityRemoved { by: account.into(), balance });
+      Ok(())
+    }
+
+    /// A call to submit the initial coin values in terms of BTC.
+    #[pallet::call_index(1)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn set_initial_price(
+      origin: OriginFor<T>,
+      prices: Prices,
+      _signature: Signature,
+    ) -> DispatchResult {
+      ensure_none(origin)?;
+
+      // set the prices
+      Oracle::<T>::set(Coin::Bitcoin, Some(prices.bitcoin));
+      Oracle::<T>::set(Coin::Monero, Some(prices.monero));
+      Oracle::<T>::set(Coin::Ether, Some(prices.ethereum));
+      Oracle::<T>::set(Coin::Dai, Some(prices.dai));
+      Ok(())
+    }
+  }
+
+  #[pallet::validate_unsigned]
+  impl<T: Config> ValidateUnsigned for Pallet<T> {
+    type Call = Call<T>;
+
+    fn validate_unsigned(_: TransactionSource, call: &Self::Call) -> TransactionValidity {
+      match call {
+        Call::set_initial_price { ref prices, ref signature } => {
+          // TODO: if this is supposed to be called after a month, serai set won't still be
+          // in the session 0? Ideally this should pull the session from Vs pallet?
+          let set = ValidatorSet { network: NetworkId::Serai, session: Session(0) };
```

Comment:
We definitely should be able to read the latest session. We won't have distinct validators since there's no staking/unstaking.